fhe-debugger/fhe-debugger-frontend/src/App.tsx

```diff
@@ -0,0 +1,233 @@
+    [onClickHandler]
+  )
+
+  const lineProps = useCallback((lineNumber: number) => {
+    return lineNumber !== selectedLine ?
+      {
+        onClick: () => onSelectLine(lineNumber),
+      } :
+      {
+        onClick: () => onSelectLine(lineNumber),
+        style: {backgroundColor: "saddlebrown"}
+      }
+  }, [onSelectLine, selectedLine]);
+
+  return (
+    <SyntaxHighlighter
+      language="rust"
+      style={atomDark}
+      showLineNumbers={true}
+      wrapLines={true}
+      useInlineStyles={true}
+      lineProps={lineProps}
+      lineNumberStyle={{minWidth: 10}}
+    >
+      {code}
+    </SyntaxHighlighter>
+  );
+};
+
+const exampleUberGraph = {
+  nodes: [
+      { id: 1, title: '2', type: 'empty' },
+      { id: 2, title: '3', type: 'empty'  },
+      { id: 3, title: '+', type: 'empty'  },
+      { id: 4, title: 'square', type: 'empty', test: 27183912  },
+  ],
+  edges: [
+    { source: 1, target: 3, directed: true, arrowhead: 'normal' },
+    { source: 2, target: 3, directed: true, arrowhead: 'normal' },
+    { source: 3, target: 4, directed: true, arrowhead: 'normal' },
+  ]
+}
+
+const exampleCode: string = `fn sudoku_proof<F: BackendField>(
+  #[constant] constraints: [[NativeField<F>; 9]; 9],
+  board: [[NativeField<F>; 9]; 9],
+) {
+  fn assert_unique_numbers<F: BackendField>(arr: [ProgramNode<NativeField<F>>; 9]) {
+      for i in 1..=9 {
+          let mut circuit = NativeField::<F>::from(1).into_program_node();
+          for a in arr {
+              circuit = circuit * (NativeField::<F>::from(i).into_program_node() - a);
+          }
+          circuit.constrain_eq(NativeField::<F>::from(0));
+      }
+  }
+  // Proves that the board matches up with the puzzle where applicable
+  let zero = NativeField::<F>::from(0).into_program_node();
+
+  for i in 0..9 {
+      for j in 0..9 {
+          let square = board[i][j].into_program_node();
+          let constraint = constraints[i][j].into_program_node();
+          (constraint * (constraint - square)).constrain_eq(zero);
+      }
+  }
+
+  // Checks rows contain every number from 1 to 9
+  for row in board {
+      assert_unique_numbers(row);
+  }
+
+  // Checks columns contain each number from 1 to 9
+  for col in 0..9 {
+      let column = board.map(|r| r[col]);
+      assert_unique_numbers(column);
+  }
+
+  // Checks squares contain each number from 1 to 9
+  for i in 0..3 {
+      for j in 0..3 {
+          let rows = &board[(i * 3)..(i * 3 + 3)];
+
+          let square = rows.iter().map(|s| &s[(j * 3)..(j * 3 + 3)]);
+
+          let flattened_sq: [ProgramNode<NativeField<F>>; 9] = square
+              .flatten()
+              .copied()
+              .collect::<Vec<_>>()
+              .try_into()
+              .unwrap_or([zero; 9]);
+
+          assert_unique_numbers(flattened_sq);
+      }
+  }
+}`
+
+const dataToGraph = (data: { graph: { graph: any; }; nodes: string | any[]; edges: string | any[]; }, incRelin: boolean) => {
```

Comment:
Make a set of types that describe the structure of the graph the frontend is giving you:

```ts
type FheProgramOperation = 'InputCiphertext' | 'Multiply' | 'Add' | ...;

type FheProgramNode = {
  operation: FheProgramOperation;
}

type Source = number;
type Target = number;

type EdgeType = 'Left' | 'Right' | ...

type FheProgramEdge = [Source, Target, EdgeType];

type FheProgramGraph = {
  nodes: FheProgramNode[];
  edges: FheProgramEdge[]
}

type FheProgram = {
  graph: { graph: { graph: FheProgramGraph }};
  data: 'Bfv'
}
```

Reply:
Then, the input type of data is FheProgram | ZkpProgram